kafka-rest/src/main/java/io/confluent/kafkarest/controllers/BrokerManagerImpl.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.kafkarest.controllers;
+
+import io.confluent.kafkarest.entities.Broker;
+import io.confluent.kafkarest.entities.Cluster;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import javax.inject.Inject;
+import javax.ws.rs.NotFoundException;
+
+final class BrokerManagerImpl implements BrokerManager {
+
+  private final ClusterManager clusterManager;
+
+  @Inject
+  BrokerManagerImpl(ClusterManager clusterManager) {
+    this.clusterManager = Objects.requireNonNull(clusterManager);
+  }
+
+  @Override
+  public CompletableFuture<List<Broker>> listBrokers(String clusterId) {
+    return clusterManager.getCluster(clusterId)
+        .thenApply(cluster ->
+            cluster.map(Cluster::getBrokers)
+                .orElseThrow(
+                    () -> new NotFoundException(
+                        String.format("Cluster %s cannot be found.", clusterId))));
```

Comment:
Is it that the cluster can't be found or no brokers. If it's the former shouldn't `clusterManager.getCluster` throw this instead?

Reply:
No cluster has been found. No brokers found is signalled by empty list. The ClusterManager signals no cluster found with Optional.empty(). Basically, empty-value for the thing you are looking for not found. Exception when you are looking for the wrong thing (e.g. for brokers of a non-existing cluster).

Reply:
[pre-note: all comments with the disclaimer that I don't know Java..]

@rnpridgeon has a good point here.

1. you're applying map to a single value, which apparently works but is not a great idea since map is something typically applied to a collection, so someone skimming it is going to pick up on the wrong signals, particularly if they don't use java because the code is just weird.
2. logically, it's more straightforward to check for the cluster existing first, and throw the exception rather than propagate this information through the map function, so you should prefer that. It's also more straightforward to write. i.e.:

```
    return clusterManager.getCluster(clusterId).thenApply(
      cluster -> cluster
        .orElseThrow(() -> new NotFoundException(String.format("Cluster %s not found.", clusterId)))
        .getBrokers());
```

further notes:

1. I'd phrase the error message differently as above.
2. indentation. i'm not sure what's javery but one thing is it's most clear if `.orElseThrow` & `.getBrokers` (or `.map` in your original version) are at the same level imo.

Reply:
I think a way to summarise what you are saying is: "get rid of the optional as soon as possible", which I completely agree. Sometimes it's not possible or convenient, but that's not the case here. You are completely correct. About the `map`, that's what you use to transform optionals in Java. I think they designed the optional [monad](https://en.wikipedia.org/wiki/Monad_(functional_programming)) as a [functor](https://en.wikipedia.org/wiki/Functor_(functional_programming)), which is why you can use `[map](https://en.wikipedia.org/wiki/Map_(higher-order_function))` on it. Regardless, I will do the change. Thanks!